napari/resources/__init__.py

```diff
@@ -1,22 +1,149 @@
+    (in this order):
+        1. The directory of *this* file (currently ``napari/resources``)
+        2. The user ~/.config/napari directory
+        3. A temporary file.
+
+    If a temporary file must be used, resources will need to be rebuilt at each
+    launch of napari (which takes ~300ms on a decent computer).
+
+    Parameters
+    ----------
+    version : str, optional
+        Version string, by default the resources will be written to a file that
+        encodes the current napari version, as well as Qt backend and version:
+        ``_qt_resources_{napari.__version__}_{API}_{QT_VERSION}.py``
+
+    overwrite : bool, optional
+        Whether to recompile and overwrite the resources.
+        By default, resources will be rebuilt if any of the following are True:
+            - the resources file does not already exist.
+            - ``overwrite`` argument is True
+            - the ``NAPARI_REBUILD_RESOURCES`` environmental variable is set
+    
+    Raises
+    ------
+    PermissionError
+        If we cannot write to any of the requested locations.
+    """
+    # the resources filename holds the napari version, Qt API, and QT version
+    version = version or f'{__version__}_{API}_{QT_VERSION}'
+    filename = f'_qt_resources_{version}.py'
+    tempfile = None
+    # see if we can write to the current napari/resources directory
+    target_file = _try_touch_file(join(abspath(dirname(__file__)), filename))
+    # if not, try to write to ~/.config/napari
+    if not target_file:
+        target_file = expanduser(join('~', '.config', 'napari', filename))
+        target_file = _try_touch_file(target_file)
+    # if that still doesn't work, create a temporary file.
+    if not target_file:
+        tempfile = NamedTemporaryFile(suffix='.py')
+        target_file = _try_touch_file(tempfile.name)
+    # if we can't even write a temporary file, we're probably screwed...
+    if not target_file:
+        raise PermissionError(
+            "Could not write qt_resources to disk. Please report this with a "
+            "description of your environment (pip freeze) at "
+            "https://github.com/napari/napari/issues"
+        )
+    overwrite = (
+        overwrite
+        or bool(environ.get('NAPARI_REBUILD_RESOURCES'))
+        or tempfile is not None
+    )
+    # this actually builds the res.qrc Qt resources file, and then from that
+    # autogenerates the python resources file that needs to be imported
+    # if the file already exists and overwrite is False, it will not be
+    # regenerated.
+    respath = build_pyqt_resources(fspath(target_file), overwrite=overwrite)
+    # import the python resources file and add to sys.modules
+    spec = spec_from_file_location("napari._qt_resources", respath)
+    module = module_from_spec(spec)
+    # important to add to sys.modules! otherwise segfault when function ends.
+    sys.modules[spec.name] = module
+    spec.loader.exec_module(module)
```

Comment:
🧙

Reply:
Oh no, thanks for the link, it's useful, I was only saying that you are a wizard. =D